 	



	<node pkg="rtabmap_ros" type="icp_odometry" name="icp_odometry" output="screen" >
 	    <remap from="scan_cloud"      to="/sensor/velodyne_bl"/>
 	    <remap from="odom"      to="/icpodom"/>
 	    <remap from="odom_info"      to="/rtabmap/odom_info"/>
	  
 	    <param name="frame_id"        type="string" value="base_link"/>   
 	    <param name="publish_tf"        type="bool" value="true"/> 
     
  	   <param name="Icp/PointToPlane"  type="string" value="false"/>
 	    <param name="Icp/VoxelSize"     type="string" value="0.05"/>
  	   <param name="Icp/Epsilon"       type="string" value="0.001"/>
 	    <param name="Icp/PointToPlaneK"  type="string" value="0"/>
  	   <param name="Icp/PointToPlaneRadius"  type="string" value="0.3"/>
  	   <param name="Icp/PointToPlaneMinComplexity"  type="string" value="0.005"/>
  	   <param name="Icp/MaxCorrespondenceDistance" type="string" value="0.1"/>
  	   <param name="Icp/PM"             type="string" value="false"/> <!-- use libpointmatcher to handle PointToPlane with 2d scans-->
  	   <param name="Icp/PMOutlierRatio" type="string" value="0.65"/><!-- 95? -->
	   <param name="Icp/RangeMin" type="string" value="0.9"/>
  	   <param name="Odom/Strategy"        type="string" value="0"/>
 	    <param name="Odom/GuessMotion"     type="string" value="true"/>
 	    <param name="Odom/ResetCountdown"  type="string" value="0"/>
  	   <param name="Odom/ScanKeyFrameThr"  type="string" value="0.9"/>
            <param name="Reg/Strategy"  type="string" value="0"/>
 	</node>
















<group ns="rtabmap">

	   <!-- Use RGBD synchronization -->
 	   <!-- Here is a general example using a standalone nodelet, 
 	        but it is recommended to attach this nodelet to nodelet 
 	        manager of the camera to avoid topic serialization -->
	   <node pkg="nodelet" type="nodelet" name="rgbd_sync" args="standalone rtabmap_ros/rgbd_sync" output="screen">
  	    <remap from="rgb/image"       to="/camera/rgb/image_rect_color"/>
  	    <remap from="depth/image"     to="/camera/depth_registered/image_raw"/>
 	    <remap from="rgb/camera_info" to="/camera/rgb/camera_info"/>
  	    <remap from="rgbd_image"      to="rgbd_image"/> <!-- output -->
      
   	   <!-- Should be true for not synchronized camera topics 
    	       (e.g., false for kinectv2, zed, realsense, true for xtion, kinect360)-->
   	   <param name="approx_sync"       value="true"/> 
  	   </node>



  	  <node name="rtabmap" pkg="rtabmap_ros" type="rtabmap" output="screen" args="--delete_db_on_start">

   	      <param name="frame_id" type="string" value="base_link"/>

   	      <param name="subscribe_depth" type="bool" value="false"/>
    	      <param name="subscribe_rgbd" type="bool" value="true"/>
    	      <param name="subscribe_scan" type="bool" value="false"/>
    	      <param name="subscribe_scan_cloud" type="bool" value="true"/>
    	      <param name="publish_tf" type="bool" value="false"/>
	
    	      <remap from="odom" to="/icpodom"/>
     	      <remap from="scan_cloud" to="/sensor/velodyne_bl"/>
    	      <remap from="rgbd_image" to="rgbd_image"/>

      	      <param name="queue_size" type="int" value="15"/>

          <!-- RTAB-Map's parameters -->
              <param name="RGBD/NeighborLinkRefining" type="string" value="true"/>
              <param name="RGBD/ProximityBySpace"     type="string" value="true"/>
       	      <param name="RGBD/AngularUpdate"        type="string" value="0.01"/>
       	      <param name="RGBD/LinearUpdate"         type="string" value="0.01"/>
       	      <param name="RGBD/OptimizeFromGraphEnd" type="string" value="false"/>
      	      <param name="Grid/FromDepth"            type="string" value="true"/> <!-- occupancy grid from lidar -->
      	      <param name="Reg/Force3DoF"             type="string" value="false"/>
              <param name="Reg/Strategy"              type="string" value="1"/> <!-- 1=ICP -->
              <param name="Kp/DetectorStrategy"    type="string" value="6"/> <!-- use same features as odom -->
          
          <!-- ICP parameters -->
  	      <param name="Icp/VoxelSize"                 type="string" value="0.05"/>
   	      <param name="Icp/MaxCorrespondenceDistance" type="string" value="0.1"/>
  	  </node>


   	  <node pkg="rtabmap_ros" type="rtabmapviz" name="rtabmapviz" args="-d $(find rtabmap_ros)/launch/config/rgbd_gui.ini" output="screen">
  	      <param name="subscribe_rgbd"      type="bool" value="true"/>
   	      <param name="subscribe_scan" type="bool" value="false"/>
 	      <param name="frame_id"            type="string" value="base_link"/>
    	      <param name="subscribe_scan_cloud" type="bool" value="true"/>
    
  	      <remap from="scan_cloud"            to="/sensor/velodyne_bl"/>

  	      <remap from="odom" to="/odometry/filtered"/>
  	      <param name="subscribe_odom_info" type="bool" value="true"/>
  	  </node>
  	</group>

  <arg name="imu_ignore_acc"          default="true" />
  <arg name="imu_remove_gravitational_acceleration" default="false" />

<node pkg="robot_localization" type="ekf_localization_node" name="ekf_localization" clear_params="true" output="screen">

      <param name="frequency" value="50"/>
      <param name="sensor_timeout" value="0.1"/>
      <param name="two_d_mode" value="false"/>

      <param name="odom_frame" value="odom"/>
      <param name="base_link_frame" value="base_link"/>
      <param name="world_frame" value="odom"/>

      <param name="transform_time_offset" value="0.0"/>

      <param name="odom0" value="/vo"/>
      <param name="imu0" value="/sensor/imu"/> 

      <!-- The order of the values is x, y, z, roll, pitch, yaw, vx, vy, vz, vroll, vpitch, vyaw, ax, ay, az. -->
      <rosparam param="odom0_config">[true, true, true,
                                      false, false, false,
                                      true, true, true,
                                      false, false, false,
                                      false, false, false]</rosparam>

      <rosparam     if="$(arg imu_ignore_acc)" param="imu0_config">[
                                     false, false, false,
                                     true,  true,  true,
                                     false, false, false,
                                     true,  true,  true,
                                     false,  false,  false] </rosparam>
      <rosparam unless="$(arg imu_ignore_acc)" param="imu0_config">[
                                     false, false, false,
                                     true,  true,  true,
                                     false, false, false,
                                     true,  true,  true,
                                     true,  true,  true] </rosparam>  
      
      <param name="odom0_differential" value="false"/>
      <param name="imu0_differential" value="false"/>

      <param name="odom0_relative" value="false"/>
      <param name="imu0_relative" value="false"/>

      <param name="imu0_remove_gravitational_acceleration" value="$(arg imu_remove_gravitational_acceleration)"/>

      <param name="print_diagnostics" value="true"/>

      <!-- ======== ADVANCED PARAMETERS ======== -->
      <param name="odom0_queue_size" value="5"/>
      <param name="imu0_queue_size" value="50"/> 

      <!-- The values are ordered as x, y, z, roll, pitch, yaw, vx, vy, vz,
           vroll, vpitch, vyaw, ax, ay, az. -->
      <rosparam param="process_noise_covariance">[0.005, 0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0.005, 0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0.006, 0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0.003, 0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0.003, 0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0.006, 0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0.0025, 0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0.0025, 0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0.004, 0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0,    0.001, 0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0.001, 0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0.002, 0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0.001, 0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0.001, 0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0.0015]</rosparam>

      <!-- The values are ordered as x, y,
           z, roll, pitch, yaw, vx, vy, vz, vroll, vpitch, vyaw, ax, ay, az. -->
           <rosparam param="initial_estimate_covariance">[1e-9, 0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    1e-9, 0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    1e-9, 0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    1e-9, 0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    1e-9, 0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    1e-9, 0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    1e-9, 0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    1e-9, 0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    1e-9, 0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    0,    1e-9,  0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     1e-9,  0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     1e-9,  0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     1e-9, 0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    1e-9, 0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    1e-9]</rosparam>

    </node>

